File: C++/JSCPP/math/trigonometry.cpp
#include <cmath>
#include <utility> // for std::pair
#include <stdexcept> // for std::invalid_argument

class Trigonometry {
public:
    // Basic Trigonometric Functions
    static double sine(double angle) {
        return std::sin(angle);
    }

    static double cosine(double angle) {
        return std::cos(angle);
    }

    static double tangent(double angle) {
        if (std::cos(angle) == 0) throw std::invalid_argument("Tangent undefined for this angle.");
        return std::tan(angle);
    }

    static double secant(double angle) {
        if (std::cos(angle) == 0) throw std::invalid_argument("Secant undefined for this angle.");
        return 1.0 / std::cos(angle);
    }

    static double cosecant(double angle) {
        if (std::sin(angle) == 0) throw std::invalid_argument("Cosecant undefined for this angle.");
        return 1.0 / std::sin(angle);
    }

    static double cotangent(double angle) {
        if (std::sin(angle) == 0) throw std::invalid_argument("Cotangent undefined for this angle.");
        return 1.0 / std::tan(angle);
    }

    // Inverse Trigonometric Functions
    static double arcsine(double value) {
        if (value < -1.0 || value > 1.0) throw std::invalid_argument("Input out of domain for arcsine.");
        return std::asin(value);
    }

    static double arccosine(double value) {
        if (value < -1.0 || value > 1.0) throw std::invalid_argument("Input out of domain for arccosine.");
        return std::acos(value);
    }

    static double arctangent(double value) {
        return std::atan(value);
    }

    // Law of Cosines
    static double law_of_cosines(double a, double b, double angle) {
        return std::sqrt(a * a + b * b - 2 * a * b * std::cos(angle));
    }

    // Law of Sines
    static double law_of_sines(double a, double A, double B) {
        if (std::sin(A) == 0) throw std::invalid_argument("Division by zero in law of sines.");
        return a * std::sin(B) / std::sin(A);
    }

    // Trigonometric Identities
    static double sine_sum(double angle1, double angle2) {
        return std::sin(angle1) * std::cos(angle2) + std::cos(angle1) * std::sin(angle2);
    }

    static double cosine_sum(double angle1, double angle2) {
        return std::cos(angle1) * std::cos(angle2) - std::sin(angle1) * std::sin(angle2);
    }

    static double tangent_sum(double angle1, double angle2) {
        double denom = 1 - std::tan(angle1) * std::tan(angle2);
        if (denom == 0) throw std::invalid_argument("Tangent sum undefined.");
        return (std::tan(angle1) + std::tan(angle2)) / denom;
    }

    static double sine_double(double angle) {
        return 2 * std::sin(angle) * std::cos(angle);
    }

    static double cosine_double(double angle) {
        return std::cos(angle) * std::cos(angle) - std::sin(angle) * std::sin(angle);
    }

    static double tangent_double(double angle) {
        double denom = 1 - std::tan(angle) * std::tan(angle);
        if (denom == 0) throw std::invalid_argument("Tangent double angle undefined.");
        return 2 * std::tan(angle) / denom;
    }

    // Coordinate Conversion
    static std::pair<double, double> polar_to_cartesian(double radius, double angle) {
        double x = radius * std::cos(angle);
        double y = radius * std::sin(angle);
        return {x, y};
    }

    static std::pair<double, double> cartesian_to_polar(double x, double y) {
        double radius = std::sqrt(x * x + y * y);
        double angle = std::atan2(y, x);
        return {radius, angle};
    }
};

File: TypeScript/JSTS/Math/Trigonometry.ts
export class Trigonometry {
    // Basic Trigonometric Functions
    static sine(angle: number): number {
        return Math.sin(angle);
    }

    static cosine(angle: number): number {
        return Math.cos(angle);
    }

    static tangent(angle: number): number {
        if (Math.cos(angle) === 0) {
            throw new Error("Tangent undefined for angles where cos(angle) = 0.");
        }
        return Math.tan(angle);
    }

    // Inverse Trigonometric Functions
    static arcsine(value: number): number {
        if (value < -1 || value > 1) {
            throw new Error("Input for arcsine must be between -1 and 1.");
        }
        return Math.asin(value);
    }

    static arccosine(value: number): number {
        if (value < -1 || value > 1) {
            throw new Error("Input for arccosine must be between -1 and 1.");
        }
        return Math.acos(value);
    }

    static arctangent(value: number): number {
        return Math.atan(value);
    }

    // Reciprocal Trigonometric Functions
    static secant(angle: number): number {
        const cos = Math.cos(angle);
        if (cos === 0) {
            throw new Error("Secant undefined for angles where cos(angle) = 0.");
        }
        return 1 / cos;
    }

    static cosecant(angle: number): number {
        const sin = Math.sin(angle);
        if (sin === 0) {
            throw new Error("Cosecant undefined for angles where sin(angle) = 0.");
        }
        return 1 / sin;
    }

    static cotangent(angle: number): number {
        const tan = Math.tan(angle);
        if (tan === 0) {
            throw new Error("Cotangent undefined for angles where tan(angle) = 0.");
        }
        return 1 / tan;
    }

    // Law of Cosines: c^2 = a^2 + b^2 - 2ab * cos(C)
    static lawOfCosines(a: number, b: number, angleC: number): number {
        return Math.sqrt(a ** 2 + b ** 2 - 2 * a * b * Math.cos(angleC));
    }

    // Law of Sines: a/sin(A) = b/sin(B) = c/sin(C)
    static lawOfSines(side: number, angle: number, otherAngle: number): number {
        return (side * Math.sin(otherAngle)) / Math.sin(angle);
    }

    // Angle Sum Formulas
    static sineSum(angleA: number, angleB: number): number {
        return Math.sin(angleA) * Math.cos(angleB) + Math.cos(angleA) * Math.sin(angleB);
    }

    static cosineSum(angleA: number, angleB: number): number {
        return Math.cos(angleA) * Math.cos(angleB) - Math.sin(angleA) * Math.sin(angleB);
    }

    static tangentSum(angleA: number, angleB: number): number {
        const denominator = 1 - Math.tan(angleA) * Math.tan(angleB);
        if (denominator === 0) {
            throw new Error("Tangent sum undefined when tan(angleA) * tan(angleB) = 1.");
        }
        return (Math.tan(angleA) + Math.tan(angleB)) / denominator;
    }

    // Double Angle Formulas
    static sineDouble(angle: number): number {
        return 2 * Math.sin(angle) * Math.cos(angle);
    }

    static cosineDouble(angle: number): number {
        return Math.cos(angle) ** 2 - Math.sin(angle) ** 2;
    }

    static tangentDouble(angle: number): number {
        const denominator = 1 - Math.tan(angle) ** 2;
        if (denominator === 0) {
            throw new Error("Tangent double-angle undefined when tan(angle)^2 = 1.");
        }
        return 2 * Math.tan(angle) / denominator;
    }

    // Coordinate Conversion
    static polarToCartesian(radius: number, angle: number): { x: number; y: number } {
        return {
            x: radius * Math.cos(angle),
            y: radius * Math.sin(angle),
        };
    }

    static cartesianToPolar(x: number, y: number): { radius: number; angle: number } {
        return {
            radius: Math.sqrt(x ** 2 + y ** 2),
            angle: Math.atan2(y, x),
        };
    }
}